#11 메뉴 추가

taekyun95 opened the pull request:
# 개요

메뉴 추가

# 이슈 티켓

#11 메뉴추가 기능 구현

# 작업 내용

- 메뉴 생성 기능 추가
  - 가게에 가게ID가 존재하는지 검증하는 기능 추가
  - 가게에 가게ID와 사징님ID가 존재하는지 검증하는 기능 추가
- 테스트용 UserDetail를 SecurityContext에 저장하는 기능 추가

bgpark82 committed:
사장님 회원가입, 식당 등록 (#9)

* #8 라이브러리 추가

* #8 git ignore 추가 (application-local.yml)

* #8 사장님 회원가입

* #8 식당 등록

* #8 사장님 회원가입 url 변경

* #8 POST 메소드 응답 상태 변경

- 204 -> 200

* #8 비밀번호 확인 로직 제거

프론트엔드에서 비밀번호 확인 처리

* #8 커스텀 예외 추가

- 예외가 계속 추가되면 @ExceptionHandler, @RestControllerAdvisor으로 변경 예정
- 참고: https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

* #8 사장님 회원가입 에지 케이스 테스트 추가

* #8 사장님 회원가입 http 요청 추가

Co-authored-by: peter-alethio <peter@alethio.io>

taekyun95 committed:
Feature/tgk 4 user curd (#7)

* #4 라이브러리 추가

spring-boot-starter-validation: 데이터 유효검 검사용
spring-boot-devtools: 개발편의성
mybatis-spring-boot-starter-test: Mybatis Test용
mapstruct: Object Mappering용

* #4 설정추가 (DB, Mybatis)

Mybatis: 
  mapper-locations: 매퍼 위치 설정
  map-underscore-to-camel-case: 테이블 컬럼명을 스네이크 케이스를 커멀케이스로 매핑해주는 설정

* #4 사용자 회원가입

* #4 UserObjectMapper 오브젝트 매핑 설정 추가

id => 무시
createDate, updateDate => LocalDateTime.now()

* #4 테스트 메소드명 한글로 변경

영어에서 한글로 변경

* #4 기능: 세션 로그인 추가

* #4 오타 수정

* #4기능: 사용자 세션 로그아웃 추가

* 코드 리뷰 반영

1. * import 풀어 사용하기
2. static String을 한 곳에서 모이기
3. 구분의 의미 없는 공백 제거
4. DTO String 속성 길이 제한
5. DateTime ​사용하는 변수에 이름을 정확히 표현
6. gradle 의존성 namespace를 그룹핑 + 사전 순 정렬

* #4 코드 리뷰 반영

1. * import 풀어 사용하기
2. static String을 한 곳에서 모이기
3. 구분의 의미 없는 공백 제거
4. DTO String 속성 길이 제한
5. DateTime ​사용하는 변수에 이름을 정확히 표현
6. gradle 의존성 namespace를 그룹핑 + 사전 순 정렬

* #4 JSON 방식 스프링 시큐리티 적용

기타
RegisterDto: @Setter 제거
UserController: 로그인, 로그아웃 제거
UserObjectMapper: password 인코딩 매핑 추가
LoginDtoValidator: 제거
JUnit Assertion => AssertJ

* 세션 저장소 변경: 로컬 저장소에서 레디스로

* 변경 LocalDateTime에서 ZonedDateTime으로

* 변경 클래스명: UserContext에서 RestUserDetails로

* #4 로그인, 로그아웃, 회원가입 URL 규칙 적용

* #4 시큐리티 클래스 명칭 수정

RestXXX => GeneralXXX

* #4 Validation 테스트 추가

* #4 AuthenticationProcessingFilter에 validation추가

* 공백 정리

taekyun95 committed:
Merge branch 'develop/user' into dev

taekyun95 committed:
Merge branch 'develop/owner' into dev

taekyun95 committed:
#11 메뉴 추가

* 추가: 메뉴 추가 기능
* 추가: 가게에 가게ID가 있는지 조회하는 기능
* 추가: 가게에 가게ID와 사장짐 ID가 있는지 조회하는 기능

taekyun95 committed:
#13 사장님 로그인 추가 (#17)

* #13 사장님 로그인 추가

* #13 OwnerAuthentication...과 GeneralAuthentication...의 공통된 부분을 추상화 처리

* 리팩토링

불필요한 파일 삭제, 패키지 이동

bgpark82 commented on src/main/java/com/flab/commerce/domain/menu/MenuController.java:
```diff
@@ -0,0 +1,50 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/stores/{storeId}/menus")
```
메뉴의 경로는 `/menus`로 해도 괜찮지 않을까요?

taekyun95 replied:
`/stores/{storeId}/menus`도 괜찮지 않나요??

bgpark82 commented on src/main/java/com/flab/commerce/domain/menu/MenuController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    if (!storeMapper.idExists(storeId)) {
+      return ResponseEntity.badRequest().body("해당 가게는 존재하지 않습니다.");
```
해당 에러들은 @RestControllerAdvisor로 해결할 수 있을 것 같습니다
나중에 같이 얘기해보시죠

f-lab-jason replied:
제 기억에도 전역으로 설정된 ExceptionHandler가 있었던 것 같은데..한군데서 처리해주는게 더 좋겠어요.

taekyun95 replied:
ExceptionHandler를 사용해보겠습니다.

bgpark82 commented on src/main/java/com/flab/commerce/domain/menu/MenuController.java:
```diff
@@ -0,0 +1,50 @@
+
+    return new ResponseEntity(HttpStatus.CREATED);
+  }
+}
```
혹시 로직을 Controller에서 다 처리한 이유가 있을까요??

taekyun95 replied:
@RestControllerAdvisor 사용하면 되는 걸 생각 못했습니다

taekyun95 replied:
유효성 검사를 말씀하시는 건가요?

bgpark82 commented on src/main/java/com/flab/commerce/domain/menu/MenuRegisterDto.java:
```diff
@@ -0,0 +1,31 @@
+  @Min(0)
+  @Max(10_000_000)
+  @NotNull
+  private Long price;
```
경계값 설정 잘 하신 것 같습니다👍
아래처럼 @Length로 통일해도 괜찮지 않을까 싶습니다

f-lab-jason replied:
`@Length`는 string 전용이라 이렇게 설정하신걸로 보여요.

f-lab-jason commented on src/main/java/com/flab/commerce/config/SecurityConfig.java:
```diff
@@ -25,6 +26,7 @@ protected void configure(HttpSecurity http) throws Exception {
         .antMatchers(HttpMethod.GET, "/users/logout").permitAll()
         .antMatchers(HttpMethod.POST, "/owners", "/owners/login").permitAll()
         .antMatchers(HttpMethod.GET, "/owners/logout").permitAll()
+        .antMatchers(HttpMethod.POST, "/store/*/menus").hasRole(Constants.OWNER)
```
Store의 endpoint도 `/stores`로 맞춰주시면 좋겠어요.

taekyun95 replied:
네 수정 알겠습니다.

f-lab-jason commented on src/main/java/com/flab/commerce/domain/menu/MenuService.java:
```diff
@@ -0,0 +1,15 @@
+
+  private final MenuMapper menuMapper;
+
+  public boolean registerMenu(Menu menu) {
```
(nit) `isMenuRegistered`

taekyun95 replied:
감사합니다

f-lab-jason commented on src/main/java/com/flab/commerce/util/Constants.java:
```diff
@@ -5,8 +5,11 @@ public class Constants {
 
   public static final String ROLE_USER = "ROLE_USER";
+
   public static final String ROLE_OWNER = "ROLE_OWNER";
 
+  public static final String OWNER = "OWNER";
```
`ROLE_OWNER`와 `OWNER`은 어떤 차이인가요?

taekyun95 replied:
hasRule에 값에 필수적으로 ROLE_ prefix가 붙는 지 알고 ROLE_OWNER에서 ROLE_ 뺀 OWNER를 추가 했습니다.
수정하겠습니다.

taekyun95 replied:
시큐리티 경로 권한 설정 hasRole에 ROLE_가 자동으로 붙어 ROLE_를 뺀 OWNER를 만들었습니다.
hasAuthority으로 변경해서 OWNER를 제거하겠습니다.

taekyun95 committed:
#11 코드 리뷰 반영

- ExceptionHandler사용
- Constants OWNER제거
- SecurityConfig "/store/*/menus" -> "/stores/*/menus"
- 과거, 사장님의가게 검증을 StoreService로 이동

taekyun95 committed:
Merge branch 'dev' into feature/11-add-menu

f-lab-jason commented on src/main/java/com/flab/commerce/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,42 @@
+
+  @ExceptionHandler
+  public ResponseEntity<Object> handleException(Exception ex) throws Exception {
+    if (ex instanceof BadInputException) {
+      return ResponseEntity.badRequest().body(ex.getMessage());
+    } else if (ex instanceof AccessDeniedException) {
+      return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(ex.getMessage());
+    } else {
+      throw ex;
+    }
```
(nit) `switch`를 쓰시는게 더 좋겠네요!

taekyun95 replied:
네 감사합니다

f-lab-jason commented on src/main/java/com/flab/commerce/domain/store/StoreService.java:
```diff
@@ -1,16 +1,28 @@
-    public boolean register(StoreRegisterDto registerRequest, Long ownerId) {
-        int countInsertRow = storeMapper.register(registerRequest.toStore(ownerId));
-        return countInsertRow == 1;
+  public boolean register(StoreRegisterDto registerRequest, Long ownerId) {
+    int countInsertRow = storeMapper.register(registerRequest.toStore(ownerId));
+    return countInsertRow == 1;
+  }
```
(nit) 보통 PK를 포함하는 register같은 행위(insert)는 void 타입으로 많이 선언합니다. 성공적으로 insert되었다면 딱히 할 일이 없을 것이고 그게 아니라면 insert시에 sql exception을 뱉을거니까 사실 mapper의 return값은 크게 신경쓰지 않습니다. exception을 뱉지 않고 countInsertRow가 0이 되는 경우는 없으니까요. (dto가 valid하다는 가정하에) 물론 boolean 리턴값을 줘도 관계없습니다 🙂

taekyun95 replied:
넵 감사합니다.

f-lab-jason commented:
'메뉴를 추가한다'는 행위는 store의 owner인지 여부와는 관계없이 가능한건가요? 삭제에서는 체크하고있는 것 같아서 따로 코멘트 남겨요
